### PR TITLE
AuthZ service: Expand the logic to also evaluate action sets

### PR DESCRIPTION
**Pull request by IevaVasiljeva**
**What is this feature?**

Expands the evaluation logic to also take into account action sets. Eg, if the request is for verb `delete` on a dashboard resource, we would evaluate the following permissions: `dashboards:delete` (the corresponding RBAC permission), `dashboards:edit` and `dashboards:admin` (the corresponding dashboard action sets), `folders:edit` and `folders:admin` (the corresponding folder action sets).

**Why do we need this feature?**

We are switching over to storing only action sets and not all the underlying permissions (the new resource permission API already stores only action sets). So it is crucial that we evaluate action sets when evaluating permissions.

**Special notes for your reviewer:**

Some design decisions:
* we could avoid having a separate `actionSet` field and have a list of `actions` instead. However, I decided to stick with a separate field for now, as I think it will make permission queries more efficient in cases where action sets do not apply (here's the [code for the query](https://github.com/grafana/grafana/blob/e273f5a608b2f9881a88b204ee144aead0aaa3f5/pkg/services/authz/rbac/store/permission_query.sql#L16-L21))
* I'm reusing the dashboard and folder action set lists defined in `ossaccesscontrol`

### CODE REVIEW

File: pkg/services/authz/rbac/service.go [hunk 1]
```diff
@@ -293,7 +294,7 @@ func (s *Service) validateListRequest(ctx context.Context, req *authzv1.ListRequ
 		return nil, err
 	}
 
-	action, err := s.validateAction(ctx, req.GetGroup(), req.GetResource(), req.GetVerb())
+	action, actionSet, err := s.validateAction(ctx, req.GetGroup(), req.GetResource(), req.GetVerb())
```
Comment (gamab) [inline pkg/services/authz/rbac/service.go]:
```suggestion
	action, actionSets, err := s.validateAction(ctx, req.GetGroup(), req.GetResource(), req.GetVerb())
```

File: pkg/services/authz/rbac/service_test.go [hunk 2]
```diff
@@ -1086,6 +1091,53 @@ func TestService_Check(t *testing.T) {
...
+				Group:     "folder.grafana.app",
+				Resource:  "folders",
+				Verb:      "delete",
+				Name:      "folder1",
+			},
+			permissions: []accesscontrol.Permission{
+				{Action: "folders:view", Scope: "folders:uid:folder1"},
+				{Action: "folders:edit", Scope: "folders:uid:other_folder"},
+			},
+			expected: false,
```
Comment (gamab) [inline pkg/services/authz/rbac/service_test.go:1139]: I don't think this test is relevant, it is in reality testing the change made further down in `(f *fakeStore) GetUserPermissions`.
The actual actionSet filtering is done by the mapper.

File: pkg/services/authz/rbac/service_test.go [hunk 3]
```diff
@@ -1824,7 +1901,13 @@ func (f *fakeStore) GetUserPermissions(ctx context.Context, namespace types.Name
 	if f.err {
 		return nil, fmt.Errorf("store error")
 	}
-	return f.userPermissions, nil
+	var permissions []accesscontrol.Permission
```
Comment (gamab) [inline pkg/services/authz/rbac/service_test.go:1904]: See previous comment, I think we can revert that change.

File: pkg/services/authz/rbac/mapper.go [hunk 4]
```diff
@@ -101,13 +112,67 @@ func newResourceTranslation(resource string, attribute string, folderSupport, sk
 	}
 }
 
+// newDashboardTranslation creates a translation for dashboards and also maps the actions to action sets
+func newDashboardTranslation(resource string, attribute string, folderSupport, skipScopeOnCreate bool) translation {
+	dashTranslation := newResourceTranslation(resource, attribute, folderSupport, skipScopeOnCreate)
```
Comment (gamab) [inline pkg/services/authz/rbac/mapper.go]: Nit. We don't need the parameters. Same for folders.
```suggestion
func newDashboardTranslation() translation {
	dashTranslation := newResourceTranslation("dashboards", "uid", true, false)
```

**Review by gamab — APPROVED** [review]
Really nice that we could integrate in the existing code so well!

Awesome work!

I couldn't find any edge case.

File: pkg/services/authz/rbac/service_test.go (hunk 2 above)
Comment (IevaVasiljeva) [reply pkg/services/authz/rbac/service_test.go:1139]: Doing an integration test felt a bit more solid to me than just testing the mapper (this way we test that the mapped action sets are actually used). But I agree that with the usage of the fake store it's still a pretty incomplete test (we know that action sets are passed to the store, but we don't test the actual store behaviour). I'm still tempted to leave it as is for now if you don't mind, and perhaps we can improve the tests in the future.

**Comment by IevaVasiljeva** [comment]
Note: the big refactor in alerting code was necessary to break the following import cycle:
```
# github.com/grafana/grafana/pkg/services/ngalert/store
package github.com/grafana/grafana/pkg/services/ngalert/store
	imports github.com/grafana/grafana/pkg/services/ngalert/testutil from alert_rule_test.go
	imports github.com/grafana/grafana/pkg/services/folder/folderimpl from testutil.go
	imports github.com/grafana/grafana/pkg/registry/apis/folders from unifiedstore.go
	imports github.com/grafana/grafana/apps/iam/pkg/reconcilers from register.go
	imports github.com/grafana/grafana/pkg/services/authz from folder_reconciler.go
	imports github.com/grafana/grafana/pkg/services/authz/rbac from rbac.go
	imports github.com/grafana/grafana/pkg/services/accesscontrol/ossaccesscontrol from mapper.go
	imports github.com/grafana/grafana/pkg/services/ngalert from receivers.go
Error: 	imports github.com/grafana/grafana/pkg/services/ngalert/store from ngalert.go: import cycle not allowed in test
```

File: pkg/services/ngalert/models/accesscontrol.go [hunk 5]
```diff
@@ -0,0 +1,43 @@
+package models
+
+import (
+	"crypto/sha1"
```
Comment (gamab) [inline pkg/services/ngalert/models/accesscontrol.go:5]:
```suggestion
	// #nosec G505 Used only for shortening the uid, not for security purposes.
	"crypto/sha1"
```

**Review by gamab — APPROVED** [review]
